### DISCUSSION TITLE
同一个collection同一个query，搜索参数也相同但结果是不一样的？

### DISCUSSION THREAD

**Discussion by gctian** [milvus-io/milvus#17860] [Q&A and General discussion]
同一个collection，同一个query向量，search_params也相同，执行两次检索，在 Topk=100时检索结果是一样的，topk=200时检索结果有约一半是不一样的，这个是什么原因呢？

index_params = {
            "metric_type":"L2",
            "index_type":"IVF_FLAT", 
            "params":{"nlist":1024} 
        } 
search_params = {"metric_type": "L2", "params": {"nprobe": 64}}

**Comment by yhmo** [comment]
1.x版本还是2.0版本？

有可能是部分IVF桶中心点离被搜索向量的距离十分接近，差别以至于小于float的精度，两次搜索选中了不同的桶。可以把搜索向量换一下再试试。如果换了不同的搜索向量仍然是这种情况，不排除是个bug。

**Reply by gctian** [reply]
2.0版本，试过不同的搜索向量，top200都是会出现不一样的结果，不过基本都是靠后的结果不一样，不知道是bug还是你说的这个情况

**Reply by yhmo** [reply]
有没有可能底库里面有部分向量完全相等？IVF在搜索的时侯是多线程计算的，如果有些向量完全像等，它们跟目标向量的距离也相等。distance相等的向量，排名的先后顺序不一定是固定的，比如1，2， 3这几条向量完全相等，它们的排名可能是132，也可能是321

**Reply by gctian** [reply]
这个有可能，或者几乎相等

**Reply by yhmo** [reply]
如果有这样的向量的话，那应该就是这个原因。

**Reply by zlw18813081566** [reply]
你好，我也遇到了类似的情况，但是我的检索结果问题很大。当topk设置10的时候，结果是正确的，设置成50之后，结果靠前的都是不相关的图片，请问是什么原因呢？

**Reply by yhmo** [reply]
> 你好，我也遇到了类似的情况，但是我的检索结果问题很大。当topk设置10的时候，结果是正确的，设置成50之后，结果靠前的都是不相关的图片，请问是什么原因呢？

topk=10拿到的10张图片，和topk=50拿到的50张图片。前者的10张是不是都在后者的50张里找得到？

**Reply by zlw18813081566** [reply]
> > 你好，我也遇到了类似的情况，但是我的检索结果问题很大。当topk设置10的时候，结果是正确的，设置成50之后，结果靠前的都是不相关的图片，请问是什么原因呢？
> 
> topk=10拿到的10张图片，和topk=50拿到的50张图片。前者的10张是不是都在后者的50张里找得到？

 不是的，不完全一样。

**Reply by yhmo** [reply]
> > > 你好，我也遇到了类似的情况，但是我的检索结果问题很大。当topk设置10的时候，结果是正确的，设置成50之后，结果靠前的都是不相关的图片，请问是什么原因呢？
> > 
> > 
> > topk=10拿到的10张图片，和topk=50拿到的50张图片。前者的10张是不是都在后者的50张里找得到？
> 
> 不是的，不完全一样。

除了top参数，其他参数都没改变的话，拿到的前10个id应该是一样的才对。除非说库里有多个id使用了相同一张图片。

**Reply by zlw18813081566** [reply]
感谢回复，根据排查id我找到问题了
